### PR TITLE
Use IsSource, IsTarget, IsRootCandidate conditions in Python FOCS parser

### PR DESCRIPTION
**Pull request by o01eg**

### CODE REVIEW

**Comment by geoffthemedio** [comment]
why does this operator get used / called?

File: parse/PythonParser.cpp [hunk 1]
```diff
@@ -170,6 +170,7 @@ PythonParser::PythonParser(PythonCommon& _python, const boost::filesystem::path&
         auto py_variable_wrapper = py::class_<variable_wrapper>("__Variable", py::no_init)
             .def(py::self_ns::self & py::other<value_ref_wrapper<double>>())
             .def(py::self_ns::self & py::other<condition_wrapper>())
+            .def(py::self_ns::self | py::other<condition_wrapper>())
```
Comment (Cjkjvfnby) [inline parse/PythonParser.cpp]: FYI you could use the python data model and define these operators via functions. 

```python
class Or:
    def __or__(self, other):
        print("Or")
        return other


assert 1 == Or() | 1
# >>> Or


# If class does not support operation, the return method of other is called.
class Nor:
    def __or__(self, other):
        print("NOR")
        return NotImplemented


class Ror:
    def __ror__(self, other):
        print("ROR")
        return other

assert Nor() | Ror()
# >>> NOR
# >>> ROR
```

**Comment by o01eg** [comment]
@geoffthemedio It is to implement next condition:

https://github.com/freeorion/freeorion/blob/bbdd9588dc3d08ac4cc0cbc8438c4e5b1813b39c/default/scripting/species/common/stealth.macros#L4-L10

To get access to `Source` fields it defined as variable, when operates with conditions or value references it casts to condition or value reference when needed.

File: parse/PythonParser.cpp (hunk 1 above)
Comment (o01eg) [reply parse/PythonParser.cpp]: Not sure if I need to do this for C++ structures

**Comment by geoffthemedio** [comment]
> To get access to Source fields it defined as variable, when operates with conditions or value references it casts to condition or value reference when needed.

I don't understand...  Where would the `|` operator get called / referenced / used between a variable and a condition?

Is it for expressions like `scope=Source() & OwnedBy(empire=Source.Owner)` that needs to alternatively interpret `Source` as either the condition (`Source()`) or as a object-value reference (`Source.Whatever`)? (without an explicit `|` in the script)

Or is there some other expression in which there's a variable like `Source.Owner` immediately followed by a condition without enclosing `( ...)` around the variable (as part of a preceeding condition definition) but with a `|` between them?

**Comment by o01eg** [comment]
> Is it for expressions like `scope=Source() & OwnedBy(empire=Source.Owner)` that needs to alternatively interpret `Source` as either the condition (`Source()`) or as a object-value reference (`Source.Whatever`)? (without an explicit `|` in the script)

@geoffthemedio Yes, for that reason. Of course it could be `scope=Source` or `scope=Source|<OR other condition`.

**Comment by geoffthemedio** [comment]
> it could be `scope=Source` or `scope=Source|<OR other condition`.

Does it:
1) See `Source` and create a Variable with `ReferenceType::SOURCE_REFERENCE` set.
2) See `|` followed by a Condition expression, and call `Variable::operator|(Condition)` to convert the Variable to a Condition (based on the ReferenceType in the Variable) and wrap it and the following Condition in an `Or` Condition?

It would be helpful to explain this with some comments.

Regardless, did you consider renaming the `Source` condition to `IsSource` in Python, and similarly for `Target` and `RootCandidate`?

**Comment by o01eg** [comment]
> Regardless, did you consider renaming the `Source` condition to `IsSource` in Python, and similarly for `Target` and `RootCandidate`?

@geoffthemedio No, suddenly I didn't get it when implementing it. Maybe better to do same instead of all those conversions.

**Comment by geoffthemedio** [comment]
I think you wanted to keep the Python format as close as possible to the Boost Spirit / C++ parser, including keeping names of stuff the same? At least in this case, simplifying the parser context dependence perhaps warrants using a different name, though...

**Comment by o01eg** [comment]
@geoffthemedio I prefer to keep same names as close as possible, but changes them in such cases. I just missed I could do it same with `Source`, `Target`, `RootCandidate` as well.

**Comment by geoffthemedio** [comment]
edit: nevermind, I was building the wrong solution...

**Pull request merged**